
## Transit Results

#### Margaret Antonio 16.08.27

### Progress

- [x] Install TRANSIT
- [x] Format data input for Transit: [WIG file format](https://genome.ucsc.edu/goldenpath/help/wiggle.html) (see link for example) which has insertion position and number of reads corresponding to that insertion position.
- [x] Run 19F Gluc (control) and Dapto (experimental condition) through [TRANSIT analysis tools](http://pythonhosted.org/tnseq-transit/transit_methods.html)
- [ ] Summarize results
- [ ] Compare TRANSIT results with MAGenTA results


## Formatting data for input

#### Normal input format

TRANSIT has an optional software for processing raw sequencing reads ([TPP - TnSeq Pre-processor](http://pythonhosted.org/tnseq-transit/tpp.html#)) and extracting insertion positions and corresponding read/template counts in WIG format for input into TRANSIT. TPP was developed for the TnSeq protocol proposed by Long et al 2015. Basic requirements for using TPP are:
- Unique barcodes for transposon junction DNA fragments (raw reads are reduced to unique template counts)
- Sequencing must be done in Paired-end (PE) mode
- Sequencing read length ~50 bp


#### Problem

We can't use TPP for our data (generated by protocol in van Opijnen et al 2015) because although TPP will accept single end reads, it expects unique barcodes for each transposon junction 

#### Solution

What's the best stage in our Tn-Seq pipeline to retrieve insertion position and template count? Ran TRANIST using a WIG file of c2 values per insertion site. Anything between alignment and calcFit stages?

